### PR TITLE
Token Queue Logging

### LINKED ISSUE #4557: Hog Factor Logging
Refined as:

- One log message when a hog group begins to be limited
- One log message when a hog group is no longer being limited
- One log message at a configurable interval, listing *all* hog groups being limited (and their total jobs)

### PR DESCRIPTION
**Pull request by cjllanwarne**
Closes #4557 

- Adds a configurable value `token-log-interval-seconds` in the `hog-safety` stanza.
- Logs when a hog group is at its limit (no more than once per `token-log-interval-seconds` seconds per hog group)
- Logs when a backend has used all tokens (no more than once per `token-log-interval-seconds` seconds per backend)
- Logs the current status of the Cromwell token queues (no more than once per `token-log-interval-seconds` seconds)

Sample log message:
```
Token Dispenser: The backend Local is starting too many jobs. New jobs are being limited.
```

Sample queue status output:
```
  "Token Dispenser state": {
    "queues": [{
      "token type": "BACKEND=Local/TOKENLIMIT=Some(10)/HOGFACTOR=5",
      "queue state": {
        "queue": [{
          "name": "4a458483",
          "queue size": 2
        }, {
          "name": "b106f1f4",
          "queue size": 2
        }],
        "pool": {
          "hog groups": [{
            "hog group": "4a458483",
            "used": 2,
            "available": false
          }, {
            "hog group": "b106f1f4",
            "used": 2,
            "available": false
          }],
          "hog limit": 2,
          "capacity": 10,
          "leased": 4
        }
      }
    }],
    "pointer": 0,
    "total token assignments": 4
  }
}
```

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala [hunk 1]
```diff
@@ -123,15 +155,39 @@ class JobExecutionTokenDispenserActor(override val serviceRegistryActor: ActorRe
...
+
+    JsObject(Map("Token Dispenser state" -> JsObject(Map(
+      "queues" -> queueStates,
+      "pointer" -> JsNumber(currentTokenQueuePointer),
+      "total token assignments" -> JsNumber(tokenAssignments.size)
+    ))))
+  }
+
+  private def logTokenAllocation(someInterval: FiniteDuration): Unit = {
+    log.info(tokenDispenserState.prettyPrint)
```
Comment (kshakir) [inline engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala]: We're slowly migrating away from spray, at a snail's pace 🐌

Can you switch this to circe? TOL we should be passing around case classes and not Json representations. But that could be a bigger fix later.

If one doesn't want to use case classes plus circe encoders, there are still ways of using [`Json.fromXyz()`](https://www.scala-exercises.org/circe/Json) to build up and pass around circe json.

File: engine/src/main/scala/cromwell/engine/workflow/tokens/TokenEventLogger.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  override def outOfTokens(backend: String): Unit = ()
+}
+
+class CachingTokenEventLogger(log: LoggingAdapter,
+                              cacheEntryTTL: FiniteDuration) extends TokenEventLogger {
+
+  log.info("Using CachingTokenEventLogger with an interval of {} to log token queue events", cacheEntryTTL)
+
+  private val cache = CacheBuilder.newBuilder()
+    .expireAfterWrite(cacheEntryTTL._1, cacheEntryTTL._2)
```
Comment (kshakir) [inline engine/src/main/scala/cromwell/engine/workflow/tokens/TokenEventLogger.scala]:
```suggestion
    .expireAfterWrite(cacheEntryTTL.length, cacheEntryTTL.unit)
```

File: engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala (hunk 1 above)
Comment (cjllanwarne) [reply engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala]: Yeah, case classes is definitely the right thing to use here, I just got lazy about defining an ADT schema first.

**Comment by cjllanwarne** [comment]
Pinging @danbills to ask whether the above sample output is sufficient for your use case or whether we need another ticket for a part II

**Comment by cjllanwarne** [comment]
@danbills that `"total token assignments": 4` represents the total number of job tokens distributed to job execution actors allowing them to start. I'm pretty sure that's what you want?

**Comment by danbills** [comment]
Yeah it is.  But this makes me think we have 2 implementations of job token managers now.

Also we should have a near-term plan for what to do in horizontal terms

File: engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala [hunk 3]
```diff
@@ -31,6 +38,27 @@ class JobExecutionTokenDispenserActor(override val serviceRegistryActor: ActorRe
...
+  }
+
+  val tokenEventLogger = effectiveLogInterval match {
+    case Some(someInterval) => new CachingTokenEventLogger(log, someInterval)
+    case None => NullTokenEventLogger
+  }
+
+  // Give the actor time to warm up, then start scheduling token allocation logging:
+  context.system.scheduler.scheduleOnce(10.seconds) {
+    effectiveLogInterval match {
```
Comment (danbills) [inline engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala]:
```
    effectiveLogInterval.foreach { someInterval =>
      log.info(s"Triggering log of token queue status every $someInterval")
      context.system.scheduler.scheduleOnce(someInterval) {
        self ! LogJobExecutionTokenAllocation(someInterval)
      }(context.dispatcher)
    }
```

File: engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala [hunk 4]
```diff
@@ -31,6 +38,27 @@ class JobExecutionTokenDispenserActor(override val serviceRegistryActor: ActorRe
     sendGaugeJob(ExecutionStatus.QueuedInCromwell.toString, tokenQueues.values.map(_.size).sum.toLong)
   }
 
+  val effectiveLogInterval = logInterval match {
+    case Some(x) if x == 0.seconds => None
+    case other => other
+  }
+
+  val tokenEventLogger = effectiveLogInterval match {
```
[the remark this answers is not in the document]
Comment (geoffjentry) [reply engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala]: Out of all various ways I've seen people unpack their `Options` it's the only one which I personally find to be completely counterintuitive and forcibly square peg/round holing.

For a period of time I was all on board with Runar's ["flat map that shit"](https://stackoverflow.com/a/8561903) tho I personally went back to the `match` fold until I saw [Odersky passionately defending it](https://www.reddit.com/r/scala/comments/449td4/fold_or_pattern_matching_with_an_option/czp154s). But really for me I think which one I prefer is more context dependent.

File: engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala [hunk 5]
```diff
@@ -63,33 +92,39 @@ class JobExecutionTokenDispenserActor(override val serviceRegistryActor: ActorRe
...
+      case Success(tokens) => tokens.toList
       case Failure(e) =>
         log.warning(s"Failed to take($n): ${e.getMessage}")
-        Iterator.empty
+        List.empty
+    }
+
+    if (nextTokens.nonEmpty) {
+      val hogGroupCounts = nextTokens.groupBy(t => t.queuePlaceholder.hogGroup).map { case (hogGroup, list) => s"$hogGroup: ${list.size}" }
+      log.info(s"Assigned new job execution tokens to the following groups: ${hogGroupCounts.mkString(", ")}")
```
[the remark this answers is not in the document]
Comment (danbills) [reply engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala]: oh I think this is post-rate limited so lgtm

File: engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala [hunk 6]
```diff
@@ -31,6 +38,23 @@ class JobExecutionTokenDispenserActor(override val serviceRegistryActor: ActorRe
     sendGaugeJob(ExecutionStatus.QueuedInCromwell.toString, tokenQueues.values.map(_.size).sum.toLong)
   }
 
+  val effectiveLogInterval: Option[FiniteDuration] = logInterval.filter(_ == 0.seconds)
```
[the remark this answers is not in the document]
Comment (danbills) [reply engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala]: oops

**Pull request merged**